server/src/main/java/org/opensearch/cluster/routing/allocation/decider/NodeOverloadAwareAllocationDecider.java

```diff
@@ -0,0 +1,119 @@
+            true, Setting.Property.Dynamic, Property.NodeScope);
+
+    private volatile int provisionedCapacity;
+
+    private volatile int overloadFactor;
+
+    private volatile boolean allowUnassignedPrimaries;
+
+    private static final Logger logger = LogManager.getLogger(NodeOverloadAwareAllocationDecider.class);
+
+    public NodeOverloadAwareAllocationDecider(Settings settings, ClusterSettings clusterSettings) {
+        this.overloadFactor = CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING.get(settings);
+        this.provisionedCapacity = CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING.get(settings);
+        this.allowUnassignedPrimaries = CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_ALLOW_UNASSIGNED_PRIMARIES_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING,
+            this::setOverloadFactor);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING,
+            this::setProvisionedCapacity);
+        clusterSettings.addSettingsUpdateConsumer(CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_ALLOW_UNASSIGNED_PRIMARIES_SETTING,
+            this::setAllowUnassignedPrimaries);
+    }
+
+    private void setAllowUnassignedPrimaries(boolean allowUnassignedPrimaries) {
+        this.allowUnassignedPrimaries = allowUnassignedPrimaries;
+    }
+
+    private void setOverloadFactor(int overloadFactor) {
+        this.overloadFactor = overloadFactor;
+    }
+
+    private void setProvisionedCapacity(int provisionedCapacity) {
+        this.provisionedCapacity = provisionedCapacity;
+    }
+
+    @Override
+    public Decision canAllocate(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return underCapacity(shardRouting, node, allocation, (count, limit) -> count >= limit);
+    }
+
+    @Override
+    public Decision canRemain(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation) {
+        return underCapacity(shardRouting, node, allocation, (count, limit) -> count > limit);
+    }
+
+    private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, RoutingAllocation allocation,
+                                   BiPredicate<Integer, Integer> decider) {
+        if (provisionedCapacity <= 0 || overloadFactor < 0 ) {
+            return allocation.decision(Decision.YES, NAME,
+                "overload awareness allocation is not enabled, set cluster setting [%s] and cluster se=tting [%s] to enable it",
+                CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_FACTOR_SETTING.getKey(),
+                CLUSTER_ROUTING_ALLOCATION_OVERLOAD_AWARENESS_PROVISIONED_CAPACITY_SETTING.getKey());
+        }
+        if (shardRouting.unassigned() && shardRouting.primary() && allowUnassignedPrimaries) {
```

Comment:
This can hurt allocation decision for ``PrimaryShardAllocator`` of ``GatewayAllocator``, if a node leaves and joins back the cluster, under an overload scenario. This can prevent primaries which were previously allocated on the nodes to go unassigned, just because node left and joined back?

I feel we should not have explicit setting to control ``allowUnassignedPrimaries``.

Reply:
I guess that decision is make by `canForceAllocatePrimary` which tries to force allocate a primary if a previous copy existed on disk?